File: ServerCore/PacketSession.cs
using System;
using Google.Protobuf;
using Google.Protobuf.Protocol;

namespace ServerCore
{
    public abstract class PacketSession : Session
    {
        readonly int packetSize = sizeof(ushort);
        public sealed override int OnReceive(ArraySegment<byte> buffer) 
        {
            int processLength = 0;

            while (true)
            {    
                if (buffer.Count < packetSize)
                {
                    break;
                }

                ushort dataSize = BitConverter.ToUInt16(buffer.Array, buffer.Offset);
                if (buffer.Count < dataSize)
                {
                    break;
                }
                OnReceivePacket(new ArraySegment<byte>(buffer.Array, buffer.Offset, dataSize));

                buffer = new ArraySegment<byte>(buffer.Array, buffer.Offset + dataSize, buffer.Count - dataSize);
                processLength += dataSize;
            }

            return processLength;
        }

        public void Send(IMessage packet)
        {
            ushort size = (ushort)packet.CalculateSize();


            string msgName = packet.Descriptor.Name.Replace("_", string.Empty);
            MsgID msgID = (MsgID)Enum.Parse(typeof(MsgID), msgName);

            ushort id = (ushort)msgID;
            byte[] sendBuffer = new byte[size + 4];
            
            Array.Copy(BitConverter.GetBytes((ushort)(size + 4)), 0, sendBuffer, 0, sizeof(ushort));
            Array.Copy(BitConverter.GetBytes(id), 0, sendBuffer, 2, sizeof(ushort));
            Array.Copy(packet.ToByteArray(), 0, sendBuffer, 4, size);
   
            Send(new ArraySegment<byte>(sendBuffer));
        }

        public abstract void OnReceivePacket(ArraySegment<byte> buffer);
    }
}

File: ServerCore/PacketSession.cpp
#include "pch.h"
#include "PacketSession.h"


//[0]                  [13][14]                 [27]
//[packet                 ][packet                 ]...
//[size(2)][id(2)][data...][size(2)][id(2)][data...]...
int PacketSession::OnRecv(BYTE* buffer, int len)
{
    int processLen = 0;
    while (true)
    {  
        int dataSize = len - processLen;

        if (dataSize < sizeof(PacketHeader))
            break; 

      
        PacketHeader header = *(PacketHeader*)(&buffer[processLen]);

        if (dataSize < header.size)
            break;

        OnRecvPacket(&buffer[processLen], header.size);

        processLen += header.size;
    }

    return processLen;
}
